server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -183,60 +183,186 @@ public CandlepinPoolManager(
+
+                            // Do some simple mapping validation. Our import method will handle minimal
+                            // population validation for us.
+                            if (content == null || content.getId() == null) {
+                                log.error("ERROR: product contains a null or incomplete product-content " +
+                                    "mapping: {}", product);
+                                throw new IllegalStateException("product contains a null or incomplete " +
+                                    "product-content mapping: " + product);
+                            }
+
+                            // We need to lock the incoming content here, but doing so will affect
+                            // the equality comparison for products. We'll correct them later.
+
+                            ContentData existingContent = contentMap.get(content.getId());
+                            if (existingContent != null && !existingContent.equals(content)) {
+                                log.warn("WARNING: Multiple versions of the same content received during " +
+                                    "refresh; discarding duplicate: {} => {}, {}",
+                                    content.getId(), existingContent, content
+                                );
+                            }
+                            else {
+                                contentMap.put(content.getId(), content);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Persist content changes
+        log.debug("Importing {} content...", contentMap.size());
+
+        // Lock our content
+        // TODO: Find a more efficient way of doing this, preferably within this method
+        for (ContentData cdata : contentMap.values()) {
+            cdata.setLocked(true);
+        }
+
+        Map<String, Content> importedContent = this.contentManager
+            .importContent(owner, contentMap, productMap.keySet());
+
+        // Determine which of our changes are net new
+        Set<Product> changedProducts = new HashSet<Product>();
+        Set<String> existingProductIds = this.ownerProductCurator
+            .filterUnknownProductIds(owner, productMap.keySet());
 
-            // Remove expired subscriptions
-            if (isExpired(sub)) {
-                deletedSubs.add(subId);
+        log.debug("Importing {} product(s)...", productMap.size());
+        Map<String, Product> importedProducts = this.productManager
+            .importProducts(owner, productMap, importedContent);
+
+        for (String pid : existingProductIds) {
+            // FIXME: This is technically wrong, at the moment. Products which did not change will
+            // still be reported as changed if they existed in the DB prior to import.
+            changedProducts.add(importedProducts.get(pid));
+        }
```

Comment:
After discussion, an RFE will be created to address this as a separate PR as this is minor.